File: src/app/api/v1/routers/backup.py
"""
백업 관리 API 라우터
"""
import zipfile
import shutil
from datetime import datetime
from pathlib import Path
from collections import defaultdict

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from src.app.infrastructure.db.session import get_db
from src.app.infrastructure.db.models import User, UserBackupSchedule, BackupHistory
from src.app.api.v1.middleware.auth import require_auth
from src.app.infrastructure.storage.local_provider import encode_username

router = APIRouter(prefix='/api/backup', tags=['backup'])

SAVE_DIR = Path('save')
BACKUP_DIR = Path('save/backups')

# 동시 백업 방지
_backup_in_progress: dict[str, bool] = {}


class MaxBackupsRequest(BaseModel):
  max_backups: int


# ─── 라우트 ─────────────────────────────────────────

@router.get('/schedule')
async def get_schedule(
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """백업 스케줄 조회"""
  result = await db.execute(
    select(UserBackupSchedule).where(UserBackupSchedule.user_id == user.username)
  )
  schedule = result.scalar_one_or_none()

  if not schedule:
    return {
      'success': True,
      'data': {
        'firstLogin': None,
        'nextBackup': None,
        'lastBackup': None,
        'maxBackups': 10,
        'isDriveUser': False,
      },
      'error': None,
    }

  return {
    'success': True,
    'data': {
      'firstLogin': schedule.first_login.isoformat() if schedule.first_login else None,
      'nextBackup': schedule.next_backup.isoformat() if schedule.next_backup else None,
      'lastBackup': schedule.last_backup.isoformat() if schedule.last_backup else None,
      'maxBackups': schedule.max_backups or 10,
      'isDriveUser': schedule.is_drive_user,
    },
    'error': None,
  }


@router.get('/status')
async def get_status(
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """백업 상태 확인"""
  count_result = await db.execute(
    select(func.count()).select_from(BackupHistory)
    .where(BackupHistory.user_id == user.username, BackupHistory.deleted_at == None)
  )
  count = count_result.scalar() or 0

  return {
    'success': True,
    'data': {
      'count': count,
      'exceeds_limit': count > 30,
      'in_progress': _backup_in_progress.get(user.username, False),
    },
    'error': None,
  }


@router.post('/run')
async def run_backup(
  confirmed: bool = Query(False),
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """수동 백업 실행"""
  if _backup_in_progress.get(user.username):
    raise HTTPException(status_code=409, detail='백업 진행 중')

  # 30개 초과 확인
  count_result = await db.execute(
    select(func.count()).select_from(BackupHistory)
    .where(BackupHistory.user_id == user.username, BackupHistory.deleted_at == None)
  )
  count = count_result.scalar() or 0

  if count >= 30 and not confirmed:
    return {
      'success': True,
      'data': {'needsConfirmation': True, 'currentCount': count},
      'error': None,
    }

  try:
    _backup_in_progress[user.username] = True

    # 사용자 디렉토리 zip 생성
    encoded = encode_username(user.username)
    user_dir = SAVE_DIR / encoded
    if not user_dir.exists():
      raise HTTPException(status_code=400, detail='백업할 데이터 없음')

    BACKUP_DIR.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
    backup_name = f'{encoded}_{timestamp}.zip'
    backup_path = BACKUP_DIR / backup_name

    with zipfile.ZipFile(str(backup_path), 'w', zipfile.ZIP_DEFLATED) as zf:
      for file_path in user_dir.rglob('*'):
        if file_path.is_file():
          arcname = file_path.relative_to(user_dir)
          zf.write(str(file_path), str(arcname))

    file_size = backup_path.stat().st_size

    # DB 기록
    history = BackupHistory(
      user_id=user.username,
      backup_path=str(backup_path),
      backup_size=file_size,
      status='completed',
      backup_location='local',
    )
    db.add(history)

    # 스케줄 갱신
    sched_result = await db.execute(
      select(UserBackupSchedule).where(UserBackupSchedule.user_id == user.username)
    )
    schedule = sched_result.scalar_one_or_none()
    if schedule:
      schedule.last_backup = datetime.utcnow()
    else:
      schedule = UserBackupSchedule(
        user_id=user.username,
        last_backup=datetime.utcnow(),
        max_backups=10,
      )
      db.add(schedule)

    await db.flush()
    return {
      'success': True,
      'data': {
        'backup_id': history.id,
        'backup_path': backup_name,
        'backup_size': file_size,
      },
      'error': None,
    }
  finally:
    _backup_in_progress.pop(user.username, None)


@router.get('/history')
async def get_history(
  limit: int = Query(20, ge=1, le=100),
  offset: int = Query(0, ge=0),
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """백업 히스토리 조회"""
  result = await db.execute(
    select(BackupHistory)
    .where(BackupHistory.user_id == user.username, BackupHistory.deleted_at == None)
    .order_by(BackupHistory.created_at.desc())
    .offset(offset).limit(limit)
  )
  backups = result.scalars().all()

  count_result = await db.execute(
    select(func.count()).select_from(BackupHistory)
    .where(BackupHistory.user_id == user.username, BackupHistory.deleted_at == None)
  )
  total = count_result.scalar() or 0

  return {
    'success': True,
    'data': {
      'backups': [
        {
          'id': b.id,
          'backup_location': b.backup_location,
          'backup_path': b.backup_path,
          'backup_size': b.backup_size,
          'created_at': b.created_at.isoformat() if b.created_at else None,
        }
        for b in backups
      ],
      'total': total,
    },
    'error': None,
  }


@router.post('/restore/{backup_id}')
async def restore_backup(
  backup_id: int,
  target_folder: str = Query(''),
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """백업에서 복원"""
  backup = await db.get(BackupHistory, backup_id)
  if not backup or backup.user_id != user.username:
    raise HTTPException(status_code=404, detail='백업 없음')

  backup_path = Path(backup.backup_path)
  if not backup_path.exists():
    raise HTTPException(status_code=404, detail='백업 파일 없음')

  # 복원 경로 검증
  encoded = encode_username(user.username)
  user_dir = SAVE_DIR / encoded
  if target_folder:
    import re
    if re.search(r'\.\.', target_folder):
      raise HTTPException(status_code=400, detail='잘못된 경로')
    restore_dir = user_dir / target_folder
  else:
    restore_dir = user_dir

  restore_dir.mkdir(parents=True, exist_ok=True)

  with zipfile.ZipFile(str(backup_path), 'r') as zf:
    zf.extractall(str(restore_dir))

  return {'success': True, 'data': {'restored': True}, 'error': None}


@router.delete('/{backup_id}')
async def delete_backup(
  backup_id: int,
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """백업 삭제 (소프트 삭제)"""
  backup = await db.get(BackupHistory, backup_id)
  if not backup or backup.user_id != user.username:
    raise HTTPException(status_code=404, detail='백업 없음')

  # 실제 파일 삭제
  backup_path = Path(backup.backup_path)
  if backup_path.exists():
    backup_path.unlink()

  backup.deleted_at = datetime.utcnow()
  await db.flush()
  return {'success': True, 'data': {'deleted': True}, 'error': None}


@router.put('/settings/max-backups')
async def set_max_backups(
  body: MaxBackupsRequest,
  user: User = Depends(require_auth),
  db: AsyncSession = Depends(get_db),
):
  """최대 백업 수 설정"""
  if body.max_backups < 1 or body.max_backups > 30:
    raise HTTPException(status_code=400, detail='1~30 사이 값 필요')

  result = await db.execute(
    select(UserBackupSchedule).where(UserBackupSchedule.user_id == user.username)
  )
  schedule = result.scalar_one_or_none()
  if not schedule:
    schedule = UserBackupSchedule(user_id=user.username, max_backups=body.max_backups)
    db.add(schedule)
  else:
    schedule.max_backups = body.max_backups

  await db.flush()
  return {'success': True, 'data': {'max_backups': body.max_backups}, 'error': None}

File: src/api/backup.js
/**
 * 백업 API 라우터
 * 백업 스케줄 조회, 수동 백업, 히스토리, 복원 기능
 */

const express = require('express');
const router = express.Router();
const fs = require('fs');
const backupService = require('../services/backupService');
const { BackupSchedule, BackupHistory } = require('../db/models/BackupSchedule');
const errorLogger = require('../services/errorLogger');

// 사용자별 백업 진행 상태 (동시 요청 방지)
const backupInProgress = new Map();

/**
 * 인증 미들웨어
 */
const requireAuth = (req, res, next) => {
  if (!req.session || !req.session.userId) {
    return res.status(401).json({ error: '로그인이 필요합니다.' });
  }
  next();
};

/**
 * 백업 스케줄 조회
 * GET /api/backup/schedule
 */
router.get('/schedule', requireAuth, async (req, res) => {
  try {
    const userId = req.session.userId;
    const schedule = await BackupSchedule.getByUserId(userId);

    if (!schedule) {
      return res.json({
        exists: false,
        message: '백업 스케줄이 없습니다.'
      });
    }

    res.json({
      exists: true,
      firstLogin: schedule.first_login,
      nextBackup: schedule.next_backup,
      lastBackup: schedule.last_backup,
      backupCount: schedule.backup_count || 0,
      isDriveUser: schedule.is_drive_user === 1,
      maxBackups: schedule.max_backups || 30
    });
  } catch (error) {
    errorLogger.error('백업 스케줄 조회 실패', error, {
      source: 'api.backup.schedule',
      userId: req.session?.userId,
      requestPath: req.originalUrl
    });
    console.error('[Backup API] Schedule error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * 백업 상태 확인 (30개 초과 여부)
 * GET /api/backup/status
 */
router.get('/status', requireAuth, (req, res) => {
  try {
    const userId = req.session.userId;
    const status = backupService.getBackupStatus(userId);

    res.json(status);
  } catch (error) {
    errorLogger.error('백업 상태 조회 실패', error, {
      source: 'api.backup.status',
      userId: req.session?.userId,
      requestPath: req.originalUrl
    });
    console.error('[Backup API] Status error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * 수동 백업 실행
 * POST /api/backup/run
 * Body: { confirmed: boolean } - 30개 초과 시 삭제 확인 여부
 */
router.post('/run', requireAuth, async (req, res) => {
  const userId = req.session.userId;

  // 동시 백업 요청 방지
  if (backupInProgress.get(userId)) {
    console.log(`[Backup API] Backup already in progress for user: ${userId}`);
    return res.status(429).json({
      success: false,
      error: '백업이 이미 진행 중입니다. 잠시 후 다시 시도해주세요.'
    });
  }

  // 백업 시작 표시
  backupInProgress.set(userId, true);

  try {
    // confirmed 값 검증 (이벤트 객체가 전달되는 경우 방지)
    const confirmedValue = req.body?.confirmed === true;
    const result = await backupService.runManualBackup(userId, confirmedValue);

    // result가 없는 경우 처리
    if (!result) {
      return res.status(500).json({
        success: false,
        error: '백업 서비스 응답 없음'
      });
    }

    // 확인이 필요한 경우
    if (result.needsConfirmation) {
      return res.json({
        success: false,
        needsConfirmation: true,
        currentCount: result.currentCount,
        maxBackups: result.maxBackups,
        oldestBackup: result.oldestBackup,
        message: result.message
      });
    }

    if (result.success) {
      const historyList = await BackupHistory.getByUserId(userId, { limit: 1 });
      const history = Array.isArray(historyList) ? historyList[0] : null;

      res.json({
        success: true,
        backupPath: history?.backup_path,
        size: history?.backup_size,
        mindmapCount: history?.mindmap_count,
        fileCount: history?.node_count,
        backupLocation: result.backupLocation || history?.backup_location || 'local',
        driveFailed: result.driveFailed || false,
        driveError: result.driveError || null
      });
    } else {
      const statusCode = result.error === '백업할 데이터가 없습니다.' ? 400 : 500;
      res.status(statusCode).json({
        success: false,
        error: result.error || '백업 실행 실패'
      });
    }
  } catch (error) {
    errorLogger.error('수동 백업 실행 실패', error, {
      source: 'api.backup.run',
      userId: req.session?.userId,
      requestPath: req.originalUrl,
      extra: { confirmed: req.body?.confirmed }
    });
    console.error('[Backup API] Run error:', error);
    res.status(500).json({ error: error.message });
  } finally {
    // 백업 완료 표시
    backupInProgress.delete(userId);
  }
});

/**
 * 백업 히스토리 조회
 * GET /api/backup/history
 */
router.get('/history', requireAuth, async (req, res) => {
  try {
    const userId = req.session.userId;
    const limit = parseInt(req.query.limit) || 10;
    const offset = parseInt(req.query.offset) || 0;

    const history = await BackupHistory.getByUserId(userId, { limit, offset });
    const totalCount = await BackupHistory.getBackupCount(userId);

    // 배열 보장
    const historyArray = Array.isArray(history) ? history : [];

    res.json({
      backups: historyArray.map(h => ({
        id: h.id,
        backupPath: h.backup_path,
        backupSize: h.backup_size,
        mindmapCount: h.mindmap_count,
        nodeCount: h.node_count,
        status: h.status,
        errorMessage: h.error_message,
        createdAt: h.created_at,
        backupLocation: h.backup_location || 'local',
        driveFileId: h.drive_file_id || null
      })),
      total: totalCount || 0,
      limit,
      offset
    });
  } catch (error) {
    errorLogger.error('백업 히스토리 조회 실패', error, {
      source: 'api.backup.history',
      userId: req.session?.userId,
      requestPath: req.originalUrl,
      extra: { limit: req.query?.limit, offset: req.query?.offset }
    });
    console.error('[Backup API] History error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * 백업에서 복원
 * POST /api/backup/restore/:backupId
 */
router.post('/restore/:backupId', requireAuth, async (req, res) => {
  try {
    const userId = req.session.userId;
    const { backupId } = req.params;
    const { targetFolder, backupFirst } = req.body;

    // 보안: targetFolder Path Traversal 방지
    if (targetFolder && (targetFolder.includes('..') || targetFolder.includes('/') || targetFolder.includes('\\'))) {
      return res.status(400).json({ error: '잘못된 폴더 경로입니다.' });
    }

    const result = await backupService.restoreFromBackup(userId, parseInt(backupId), targetFolder, { backupFirst: backupFirst !== false });

    res.json(result);
  } catch (error) {
    errorLogger.error('백업 복원 실패', error, {
      source: 'api.backup.restore',
      userId: req.session?.userId,
      requestPath: req.originalUrl,
      extra: { backupId: req.params?.backupId, targetFolder: req.body?.targetFolder }
    });
    console.error('[Backup API] Restore error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * 백업 삭제 (실제 파일 삭제 + DB 소프트 삭제)
 * DELETE /api/backup/:backupId
 * Drive 백업: Drive에서 실제 파일 삭제
 * 로컬 백업: 디스크에서 실제 ZIP 파일 삭제
 */
router.delete('/:backupId', requireAuth, async (req, res) => {
  try {
    const userId = req.session.userId;
    const { backupId } = req.params;

    // 백업 소유권 확인
    const backup = await BackupHistory.getById(parseInt(backupId));
    if (!backup || backup.user_id !== userId) {
      return res.status(404).json({ error: '백업을 찾을 수 없습니다.' });
    }

    // 실제 파일 삭제
    if (backup.backup_location === 'drive' && backup.drive_file_id) {
      // Drive 백업: 클라우드에서 실제 삭제
      try {
        const driveService = require('../services/driveService');
        await driveService.deleteBackup(userId, backup.drive_file_id);
        console.log(`[Backup API] Drive 파일 삭제 완료: fileId=${backup.drive_file_id}`);
      } catch (driveErr) {
        // Drive 삭제 실패해도 DB 삭제는 진행 (파일이 이미 없을 수 있음)
        console.warn(`[Backup API] Drive 파일 삭제 실패 (DB 삭제 계속): ${driveErr.message}`);
      }
    } else if (backup.backup_path) {
      // 로컬 백업: 디스크에서 실제 삭제
      try {
        if (fs.existsSync(backup.backup_path)) {
          if (backup.backup_path.endsWith('.zip')) {
            fs.unlinkSync(backup.backup_path);
          } else {
            fs.rmSync(backup.backup_path, { recursive: true, force: true });
          }
          console.log(`[Backup API] 로컬 파일 삭제 완료: ${backup.backup_path}`);
        }
      } catch (fsErr) {
        console.warn(`[Backup API] 로컬 파일 삭제 실패 (DB 삭제 계속): ${fsErr.message}`);
      }
    }

    // DB 소프트 삭제
    const result = await BackupHistory.softDelete(parseInt(backupId));
    if (!result || result.changes === 0) {
      return res.status(500).json({ error: '백업 삭제에 실패했습니다.' });
    }

    res.json({ success: true, message: '백업이 삭제되었습니다.' });
  } catch (error) {
    errorLogger.error('백업 삭제 실패', error, {
      source: 'api.backup.delete',
      userId: req.session?.userId,
      requestPath: req.originalUrl,
      extra: { backupId: req.params?.backupId }
    });
    console.error('[Backup API] Delete error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * 최대 백업 수 설정
 * PUT /api/backup/settings/max-backups
 */
router.put('/settings/max-backups', requireAuth, async (req, res) => {
  try {
    const userId = req.session.userId;
    const { maxBackups } = req.body;

    if (!maxBackups || maxBackups < 1 || maxBackups > 30) {
      return res.status(400).json({ error: '최대 백업 수는 1-30 사이여야 합니다.' });
    }

    await BackupSchedule.setMaxBackups(userId, maxBackups);

    res.json({ success: true, maxBackups });
  } catch (error) {
    errorLogger.error('백업 설정 변경 실패', error, {
      source: 'api.backup.settings',
      userId: req.session?.userId,
      requestPath: req.originalUrl,
      extra: { maxBackups: req.body?.maxBackups }
    });
    console.error('[Backup API] Settings error:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
